### PR TITLE
stmhal/modmachine:  New algorithm to find PLL values.

### PR DESCRIPTION
**Pull request by Anton-2**
This one works for 25MHz HSE like on the STM32F7DISC.

We find the best suitable vco_out first, keeping USB at 48MHz.
Then we solve m and n by iterating over n, and calculating m, once.

It gives the same m, n, p, q as the previous algorithm for  HSE_VALUE = 8MHz, and should work with any HSE_VALUE.

### CODE REVIEW

File: stmhal/modmachine.c [hunk 1]
```diff
@@ -207,49 +207,49 @@ STATIC mp_obj_t machine_freq(mp_uint_t n_args, const mp_obj_t *args) {
...
-            for (; wanted_sysclk > 0; wanted_sysclk--) {
+            // sysclk = (HSE_VALUE * n) / (m * p). pll48ck =  (HSE_VALUE * n) / (m * q). We want pll48ck at 48Mhz
+            // valid sysclk and p found at or below wanted_sysclk
+            uint32_t target_sysclk = 0, target_p = 0;
+            for (q = 2; q <= 15; q++) {
+                // compute VCO_OUT with pll48ck at 48Mhz
+                uint32_t vco_out = 48 * q;
+                // make sure VCO_OUT is between 192MHz and 432MHz
+                if (vco_out < 192 || vco_out > 432) {
+                    continue;
```
Comment (dhylands) [inline stmhal/modmachine.c]: It seems that if you range q from 4 to 9 then vco_out will always be between 192 and 432 and you can remove this test.

**Comment by Anton-2** [comment]
I'll update this PR as suggested by @dhylands once I'm done summarizing the constraints difference between various F4 and F7 devices.

**Comment by Anton-2** [comment]
According to datasheets, all F4/F7 MCU except STM32F401 have VCO min=100Mhz and N min=50.
But current version of stmhal in micropython checks that N is >= 192.

So, as per  @dhylands suggestion, we're sweeping q from 4 to 9, so vco_out is between 192 and 432 MHz.
I've added a test to check that m is between 2 and 63.

Not trying q=3 and n between 50..192 prevent us to reach 18MHz, a

**Comment by dpgeorge** [comment]
Thanks for attempting this.  But I think it's better to keep the current algorithm and improve it so that N/M doesn't need to be an integer.  The reason why it's restricted to an int in the current algo is just for simplicity, so that the code can work without using floats.  It is possible to rewrite just the M-searching loop to not use float and have N/M not being an int.

Your algo seems to be slower in that it searches all combinations of P and Q (4*6=24 combos), and then does a search over N (possibly up to 240 different values).  My original version searches just over P (4 values) and then scans M (max 62 values).

**Comment by Anton-2** [comment]
Searching for all P and Q was done to avoid looping ```wanted_sysclk``` down until a correct value is found when the user gives an invalid frequency, but I agree that it's not the common case.

I've got a proof of concept working of searching M directly using fixed point arithmetic, I will make it a proper PR this evening, after testing it on a real board.

**Comment by dpgeorge** [comment]
> Searching for all P and Q was done to avoid looping wanted_sysclk down until a correct value is found when the user gives an invalid frequency, but I agree that it's not the common case.

True, yes, I forgot about that outer-most loop in my implementation.  Well, as you say, the common case would be setting the freq to a value that is known to work, so let's optimise for that.

> I've got a proof of concept working of searching M directly using fixed point arithmetic, I will make it a proper PR this evening, after testing it on a real board.

Sounds good.  Actually, what I have locally is a Python script that implements the same algo and which I use for testing.  Since this script will be useful for testing I pushed it in 0fd3d8d19fe63e1ddbe6bcf86efef53406c21844.  Please update that script with the new algo that you implement.

**Comment by dpgeorge** [comment]
In 41b4686dd75fde815cf9c12593f223b4305e54f9 the PLL frequency selection is changed to use a simple look-up-table in the C code.  This table is generated by the script boards/pllvalues.py at build time.  The advantages of this approach (over computing the PLL values at run-time) are simpler, smaller and faster code, and more flexibility in selecing valid PLL values for a particular board.  In particular, boards with a 25MHz HSE can now have their frequency changed at run time. 

So this PR here is now obsolete.

@Anton-2 feel free to improve the pllvalues.py script to be able to find more frequencies (eg by relaxing the constraints on VCO_OUT and/or N) and submit a new PR if you like.

**Pull request closed without merging**